node/src/components/chain_synchronizer/operations.rs

```diff
@@ -341,125 +347,196 @@ impl CanUseSyncingNodes for Deploy {}
 impl CanUseSyncingNodes for BlockAndDeploys {}
 impl CanUseSyncingNodes for BlockHeadersBatch {}
 
-/// Returns fully-connected, non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_non_syncing_peers<REv>(
-    ctx: &ChainSyncContext<'_, REv>,
-) -> Vec<NodeId>
+/// Gets a list of peers suitable for the fetch operation.
+async fn get_peers<REv>(include_syncing: bool, ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
 where
     REv: From<NetworkInfoRequest>,
 {
-    let mut peer_list = ctx
-        .effect_builder
-        .get_fully_connected_non_syncing_peers()
-        .await;
+    let mut peer_list = if include_syncing {
+        ctx.effect_builder.get_fully_connected_peers().await
+    } else {
+        ctx.effect_builder
+            .get_fully_connected_non_syncing_peers()
+            .await
+    };
     ctx.filter_bad_peers(&mut peer_list);
     peer_list
 }
 
-/// Returns fully-connected, syncing and non-syncing peers that are known to be not banned.
-async fn get_filtered_fully_connected_peers<REv>(ctx: &ChainSyncContext<'_, REv>) -> Vec<NodeId>
+/// Fetching should stop only when the network is connected and all retries have been exhausted.
+fn should_stop_fetching(
+    retry_count: &mut usize,
+    has_connected_to_network: bool,
+    max_sync_fetch_attempts: usize,
+) -> bool {
+    if !has_connected_to_network {
+        return false;
+    }
+
+    *retry_count += 1;
+    if *retry_count <= max_sync_fetch_attempts {
```

Comment:
This could be

```rust
if *retry_count <= max_sync_fetch_attempts {
    true
} else {
    false
}
```

or my preference just

```rust
    // ...
    *retry_count += 1;
    *retry_count <= max_sync_fetch_attempts
}
```

Reply:
The "other way round" part clearly shows that an explicit `if` isn't that bad sometimes :-)

Reply:
In the UK, that code is correct though, right? One drives on the other side of the road, so I assume you also put the operands on the other side of the operators?

Reply:
That code will still fail in the UK I'm afraid.
Anyway, we agreed to inline the `should_stop_fetching` and get rid of the tests for the sake of clarity of `fetch_with_retries`.

The change is here: https://github.com/casper-network/casper-node/pull/3279/commits/bc168bc70f2f3c3d85a001a36556297be734e1db